File: src/test/java/com/lalkalol/mujahid/commands/impl/PlaybackCommandsTest.java
package com.lalkalol.mujahid.commands.impl;

import com.lalkalol.mujahid.audio.AudioFilter;
import com.lalkalol.mujahid.audio.LoopMode;
import com.lalkalol.mujahid.commands.CommandTestFixtures;
import dev.arbjerg.lavalink.client.Link;
import dev.arbjerg.lavalink.client.player.PlayerUpdateBuilder;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

class PlaybackCommandsTest {

    @Test
    void loopCommand_setsModeAndReplies() {
        var h = CommandTestFixtures.harness().withStringOption("mode", "queue");

        new LoopCommand().execute(h.context());

        verify(h.scheduler).setLoopMode(LoopMode.QUEUE);
        assertTrue(h.lastReplyDescription().contains("queue"));
    }

    @Test
    void clearCommand_rejectsEmptyQueue() {
        var h = CommandTestFixtures.harness();
        when(h.scheduler.queueSize()).thenReturn(0);

        new ClearCommand().execute(h.context());

        assertTrue(h.lastEphemeralDescription().contains("already empty"));
    }

    @Test
    void clearCommand_clearsQueueAndReportsCount() {
        var h = CommandTestFixtures.harness();
        when(h.scheduler.queueSize()).thenReturn(3);

        new ClearCommand().execute(h.context());

        verify(h.scheduler).clearQueue();
        assertTrue(h.lastReplyDescription().contains("3"));
    }

    @Test
    void removeCommand_reportsMissingPosition() {
        var h = CommandTestFixtures.harness().withIntOption("position", 5);
        when(h.scheduler.removeAt(4)).thenReturn(null);

        new RemoveCommand().execute(h.context());

        assertTrue(h.lastEphemeralDescription().contains("position 5"));
    }

    @Test
    void removeCommand_removesTrackByTitle() {
        var h = CommandTestFixtures.harness().withIntOption("position", 1);
        var removed = h.mockTrack("Cool Song");
        when(h.scheduler.removeAt(0)).thenReturn(removed);

        new RemoveCommand().execute(h.context());

        assertTrue(h.lastReplyDescription().contains("Cool Song"));
    }

    @Test
    void volumeCommand_requiresActivePlayer() {
        var h = CommandTestFixtures.harness()
                .withIntOption("level", 80)
                .withNoActivePlayer();

        new VolumeCommand().execute(h.context());

        assertTrue(h.lastEphemeralDescription().contains("Nothing is playing"));
    }

    @Test
    void volumeCommand_clampsAndAppliesLevel() {
        var h = CommandTestFixtures.harness()
                .withIntOption("level", 250)
                .withActivePlayer();

        new VolumeCommand().execute(h.context());

        verify(h.scheduler).applyVolume(200);
        assertTrue(h.lastReplyDescription().contains("200%"));
    }

    @Test
    void pauseCommand_rejectsWhenAlreadyPaused() {
        var h = CommandTestFixtures.harness().withPausedPlayer();

        new PauseCommand().execute(h.context());

        assertTrue(h.lastEphemeralDescription().contains("Already paused"));
    }

    @Test
    void pauseCommand_pausesActivePlayback() {
        var h = CommandTestFixtures.harness().withPlayingPlayer();
        Link link = mock(Link.class);
        PlayerUpdateBuilder builder = mock(PlayerUpdateBuilder.class);
        when(h.musicManager.getCachedLink()).thenReturn(link);
        when(link.createOrUpdatePlayer()).thenReturn(builder);
        when(builder.setPaused(true)).thenReturn(builder);

        new PauseCommand().execute(h.context());

        verify(builder).setPaused(true);
        assertEquals("Paused.", h.lastReplyDescription());
    }

    @Test
    void filterCommand_clearsWhenOffSelected() {
        var h = CommandTestFixtures.harness()
                .withStringOption("preset", "off")
                .withActivePlayer();

        new FilterCommand().execute(h.context());

        verify(h.scheduler).applyFilter(AudioFilter.OFF);
        assertEquals("Filters cleared.", h.lastReplyDescription());
    }

    @Test
    void filterCommand_appliesNamedPreset() {
        var h = CommandTestFixtures.harness()
                .withStringOption("preset", "nightcore")
                .withActivePlayer();

        new FilterCommand().execute(h.context());

        verify(h.scheduler).applyFilter(AudioFilter.NIGHTCORE);
        assertTrue(h.lastReplyDescription().contains("Nightcore"));
    }

    @Test
    void skipCommand_requiresConnectionAndActivePlayer() {
        var h = CommandTestFixtures.harness()
                .memberNotInVoice()
                .withActivePlayer();

        new SkipCommand().execute(h.context());

        assertTrue(h.lastEphemeralDescription().contains("voice channel"));
    }

    @Test
    void skipCommand_reportsEmptyQueueAfterSkip() {
        var h = CommandTestFixtures.harness()
                .bothInVoice(1L, "General")
                .withActivePlayer();
        when(h.scheduler.skip()).thenReturn(null);

        new SkipCommand().execute(h.context());

        assertTrue(h.lastReplyDescription().contains("queue is now empty"));
    }

    @Test
    void skipCommand_reportsNextTrack() {
        var h = CommandTestFixtures.harness()
                .bothInVoice(1L, "General")
                .withActivePlayer();
        var next = h.mockTrack("Next Up");
        when(h.scheduler.skip()).thenReturn(next);

        new SkipCommand().execute(h.context());

        assertTrue(h.lastReplyDescription().contains("Next Up"));
    }
}

File: src/test/kotlin/com/lalkalol/mujahid/commands/impl/PlaybackCommandsTest.kt
package com.lalkalol.mujahid.commands.impl

import com.lalkalol.mujahid.audio.AudioFilter
import com.lalkalol.mujahid.audio.LoopMode
import com.lalkalol.mujahid.commands.CommandTestFixtures
import dev.arbjerg.lavalink.client.Link
import dev.arbjerg.lavalink.client.player.PlayerUpdateBuilder
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever

class PlaybackCommandsTest {
    @Test
    fun loopCommand_setsModeAndReplies() = runBlocking {
        val h = CommandTestFixtures.harness().withStringOption("mode", "queue")

        LoopCommand().execute(h.context())

        verify(h.scheduler).setLoopMode(LoopMode.QUEUE)
        assertTrue(h.lastReplyDescription().contains("queue"))
    }

    @Test
    fun clearCommand_rejectsEmptyQueue() = runBlocking {
        val h = CommandTestFixtures.harness()
        whenever(h.scheduler.queueSize()).thenReturn(0)

        ClearCommand().execute(h.context())

        assertTrue(h.lastEphemeralDescription().contains("already empty"))
    }

    @Test
    fun clearCommand_clearsQueueAndReportsCount() = runBlocking {
        val h = CommandTestFixtures.harness()
        whenever(h.scheduler.queueSize()).thenReturn(3)

        ClearCommand().execute(h.context())

        verify(h.scheduler).clearQueue()
        assertTrue(h.lastReplyDescription().contains("3"))
    }

    @Test
    fun removeCommand_reportsMissingPosition() = runBlocking {
        val h = CommandTestFixtures.harness().withIntOption("position", 5)
        whenever(h.scheduler.removeAt(4)).thenReturn(null)

        RemoveCommand().execute(h.context())

        assertTrue(h.lastEphemeralDescription().contains("position 5"))
    }

    @Test
    fun removeCommand_removesTrackByTitle() = runBlocking {
        val h = CommandTestFixtures.harness().withIntOption("position", 1)
        val removed = h.mockTrack("Cool Song")
        whenever(h.scheduler.removeAt(0)).thenReturn(removed)

        RemoveCommand().execute(h.context())

        assertTrue(h.lastReplyDescription().contains("Cool Song"))
    }

    @Test
    fun volumeCommand_requiresActivePlayer() = runBlocking {
        val h = CommandTestFixtures.harness()
            .withIntOption("level", 80)
            .withNoActivePlayer()

        VolumeCommand().execute(h.context())

        assertTrue(h.lastEphemeralDescription().contains("Nothing is playing"))
    }

    @Test
    fun volumeCommand_clampsAndAppliesLevel() = runBlocking {
        val h = CommandTestFixtures.harness()
            .withIntOption("level", 250)
            .withActivePlayer()

        VolumeCommand().execute(h.context())

        verify(h.scheduler).applyVolume(200)
        assertTrue(h.lastReplyDescription().contains("200%"))
    }

    @Test
    fun pauseCommand_rejectsWhenAlreadyPaused() = runBlocking {
        val h = CommandTestFixtures.harness().withPausedPlayer()

        PauseCommand().execute(h.context())

        assertTrue(h.lastEphemeralDescription().contains("Already paused"))
    }

    @Test
    fun pauseCommand_pausesActivePlayback() = runBlocking {
        val h = CommandTestFixtures.harness().withPlayingPlayer()
        val link = mock<Link>()
        val builder = mock<PlayerUpdateBuilder>()
        whenever(h.musicManager.getCachedLink()).thenReturn(link)
        whenever(link.createOrUpdatePlayer()).thenReturn(builder)
        whenever(builder.setPaused(true)).thenReturn(builder)

        PauseCommand().execute(h.context())

        verify(builder).setPaused(true)
        assertEquals("Paused.", h.lastReplyDescription())
    }

    @Test
    fun filterCommand_clearsWhenOffSelected() = runBlocking {
        val h = CommandTestFixtures.harness()
            .withStringOption("preset", "off")
            .withActivePlayer()

        FilterCommand().execute(h.context())

        verify(h.scheduler).applyFilter(AudioFilter.OFF)
        assertEquals("Filters cleared.", h.lastReplyDescription())
    }

    @Test
    fun filterCommand_appliesNamedPreset() = runBlocking {
        val h = CommandTestFixtures.harness()
            .withStringOption("preset", "nightcore")
            .withActivePlayer()

        FilterCommand().execute(h.context())

        verify(h.scheduler).applyFilter(AudioFilter.NIGHTCORE)
        assertTrue(h.lastReplyDescription().contains("Nightcore"))
    }

    @Test
    fun skipCommand_requiresConnectionAndActivePlayer() = runBlocking {
        val h = CommandTestFixtures.harness()
            .memberNotInVoice()
            .withActivePlayer()

        SkipCommand().execute(h.context())

        assertTrue(h.lastEphemeralDescription().contains("voice channel"))
    }

    @Test
    fun skipCommand_reportsEmptyQueueAfterSkip() = runBlocking {
        val h = CommandTestFixtures.harness()
            .bothInVoice(1L, "General")
            .withActivePlayer()
        whenever(h.scheduler.skip()).thenReturn(null)

        SkipCommand().execute(h.context())

        assertTrue(h.lastReplyDescription().contains("queue is now empty"))
    }

    @Test
    fun skipCommand_reportsNextTrack() = runBlocking {
        val h = CommandTestFixtures.harness()
            .bothInVoice(1L, "General")
            .withActivePlayer()
        val next = h.mockTrack("Next Up")
        whenever(h.scheduler.skip()).thenReturn(next)

        SkipCommand().execute(h.context())

        assertTrue(h.lastReplyDescription().contains("Next Up"))
    }
}
